src/Storage/StreamWrapper.php

```diff
@@ -0,0 +1,580 @@
+<?php
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage;
+
+use Google\Cloud\Exception\NotFoundException;
+use Google\Cloud\Exception\ServiceException;
+use GuzzleHttp\Psr7\CachingStream;
+use GuzzleHttp\Psr7;
+
+/**
+ * A streamWrapper implementation for handling `gs://bucket/path/to/file.jpg`.
+ * Note that you can only open a file with mode 'r', 'rb', 'rb', 'w', 'wb', or 'wt'.
+ *
+ * See: http://php.net/manual/en/class.streamwrapper.php
+ */
+class StreamWrapper
+{
+    const DEFAULT_PROTOCOL = 'gs';
+
+    const FILE_WRITABLE_MODE = 33206; // 100666 in octal
+    const FILE_READABLE_MODE = 33060; // 100444 in octal
+    const DIRECTORY_WRITABLE_MODE = 16895; // 40777 in octal
+    const DIRECTORY_READABLE_MODE = 16676; // 40444 in octal
+
+    // Must be public according to the PHP documentation
+    public $context;
+
+    // a GuzzleHttp\Psr7\StreamInterface instance
+    private $stream;
+
+    private $protocol;
+    private $bucket;
+    private $file;
```

Comment:
These types should be documented with `@var`. If they're nullable, use `<type>|null`.